docs/dev/guidelines/server.rst

```diff
@@ -350,3 +350,99 @@ Please read `the AssertJ documentation <https://assertj.github.io/doc/#assertj-c
+https://www.baeldung.com/spring-tests
+
+If you want to write tests for Programming Exercises to test student's submissions check out `this <https://confluence.ase.in.tum.de/display/ArTEMiS/Best+Practices+for+writing+Java+Programming+Exercise+Tests+in+Artemis>`__.
+
+
+23. Avoid using @MockBean
+=========================
+
+Do not use the ``@SpyBean`` or ``@MockBean`` annotation unless absolutely necessary, or possibly in an abstract Superclass. If you want to see more in detail why, take a look `here <https://www.baeldung.com/spring-tests>`__.
+Basically, every time ``@MockBean`` appears in a class, the ApplicationContext cache gets marked as dirty, hence the runner will clean the cache after the test-class is done and restarts the application context.
+This leads to a large overhead, which tends to make the tests take a lot more time.
+
+Here is an example how to replace a ``@SpyBean``. We wanted to test a catch case which is only executed if an IOException gets thrown. We did this by mocking the service method and making it throw an Exception.
+
+.. code-block:: java
+
+    public class TestExport extends AbstractSpringIntegrationBambooBitbucketJiraTest {
+        @SpyBean
+        private FileUploadSubmissionExportService fileUploadSubmissionExportService;
+
+        @Test
+        @WithMockUser(username = "instructor1", roles = "INSTRUCTOR")
+        public void testExportAll_IOException() throws Exception {
+            doThrow(IOException.class).when(fileUploadSubmissionExportService).export(any(), any());
+            request.postWithResponseBodyFile("/api/file-upload-export/" + fileUploadExercise.getId()", HttpStatus.BAD_REQUEST);
+        }
+    }
+
+As mentioned above, we should really avoid this
+Instead we can use `Static Mocks <https://asolntsev.github.io/en/2020/07/11/mockito-static-methods/>`_. When we look deeper in the ``export()`` method we find that there is a call of ``File.newOutputStream(..)``.
+Now, instead of mocking the whole Service we can just mock this static method, like this:
+
+.. code-block:: java
+
+    public class TestExport extends AbstractSpringIntegrationBambooBitbucketJiraTest {
+        // No beans used anymore
+        @Test
+        @WithMockUser(username = "instructor1", roles = "INSTRUCTOR")
+        public void testExportAll_IOException() throws Exception {
+            MockedStatic<Files> mockedFiles = mockStatic(Files.class);
+            mockedFiles.when(() -> Files.newOutputStream(any(), any())).thenThrow(IOException.class);
+            request.postWithResponseBodyFile("/api/file-upload-export/" + fileUploadExercise.getId()", HttpStatus.BAD_REQUEST);
+
+            mockedFiles.close();
+        }
+    }
+
+You should notice here that we can avoid the use of a Bean and also test deeper. Instead of mocking the uppermost method we will only throw the exception at the place where it could actually happen. Very important to mention is that you need to close the mock at the end of the test again.
```

Comment:
```suggestion
You should notice here that we can avoid the use of a Bean and also test deeper. Instead of mocking the uppermost method we only throw the exception at the place where it could actually happen. Very important to mention is that you need to close the mock at the end of the test again.
```

Reply:
The `will` doesn't feel wrong here to me, but it's a matter of taste I suppose :P